test/store.unittest.ts

```diff
@@ -0,0 +1,757 @@
+import {ScopedConfig, Store} from '../src/Store';
+import {RequiredRecursive} from '../src/ConfigUtil';
+import {MaskWatch, ScopeWatch} from '../src/Watch';
+
+import {Scope} from '../src/Types';
+
+import {addConsoleSpies, ConsoleSpy} from './utils/console';
+
+/**
+ * Configuration object used for testing.
+ *
+ * In real usage, this would come from a JSON schema.
+ */
+interface Config {
+    // Generic properties of each primitive type
+    bool?: boolean;
+    num?: number;
+    str?: string;
+
+    // Object type, where all members are optional (borrowing the 'features' use-case from layouts config)
+    features?: {
+        snap?: boolean;
+        dock?: boolean;
+        tab?: boolean;
+    };
+
+    // Object type, must either be fully-specified or not at all specified (borrowing the 'tabstrip' use-case from layouts config)
+    tabstrip?: {
+        url: string;
+        height: number;
+    };
+
+    // Object type, with some more detailed properties
+    theme?: {
+        scheme?: string;
+        opacity?: number;
+        backColour?: number|string;
+        borderColour?: number|string;
+        borderThickness?: number;
+    };
+}
+
+/**
+ * To reduce verbosity of tests, pre-define scopes that can be used for defining rules and querying the store.
+ */
+const scopes = {
+    // "Generic" scopes with no contextual parameters
+    service: {level: 'service'} as Scope,
+    desktop: {level: 'desktop'} as Scope,
+
+    // Three dummy applications
+    app_1: {level: 'application', uuid: 'app-1'} as Scope,
+    app_2: {level: 'application', uuid: 'app-2'} as Scope,
+    app_3: {level: 'application', uuid: 'app-3'} as Scope,
+
+    // Define two windows for each application
+    win_1_1: {level: 'window', uuid: 'app-1', name: 'win-1'} as Scope,
+    win_1_2: {level: 'window', uuid: 'app-1', name: 'win-2'} as Scope,
+    win_2_1: {level: 'window', uuid: 'app-2', name: 'win-1'} as Scope,
+    win_2_2: {level: 'window', uuid: 'app-2', name: 'win-2'} as Scope,
+    win_3_1: {level: 'window', uuid: 'app-3', name: 'win-1'} as Scope,
+    win_3_2: {level: 'window', uuid: 'app-3', name: 'win-2'} as Scope
+};
+
+/**
+ * Parameters of the callbacks added to the `onAdd` and `onRemove` signals of a `Watch` object.
+ */
+type WatchCallbackParams = [ScopedConfig<Config>, Scope];
+
+let store: Store<Config>;
+const defaults: RequiredRecursive<Config> = {
+    bool: true,
+    num: 0,
+    str: '',
+    features: {snap: true, dock: true, tab: true},
+    tabstrip: {url: 'http://localhost/tabstrip.html', height: 60},
+    theme: {
+        scheme: 'default',
+        opacity: 1,
+        backColour: 0x808080,
+        borderColour: 'black',
+        borderThickness: 1
+    }
+};
+
+let spyConsole: ConsoleSpy;
+
+beforeEach(() => {
+    spyConsole = addConsoleSpies();
+
+    jest.restoreAllMocks();
+    // @ts-ignore Complex types cant be represented.  All is working.
+    store = new Store(defaults);
```

Comment:
Does adding `<Config>` fix this instance of that error too?

Reply:
Yes it fixed it.  I think its my TS version (3.5.2) causing these issues.